src/vptranslator.rs

```diff
@@ -0,0 +1,253 @@
+            });
+        }
+    }
+}
+
+fn fill_operating_days(week: u8, operating_days: &mut Vec<Weekday>) {
+    for i in 0..7 {
+        if week & (1 << i) == (1 << i) {
+            operating_days.push(Weekday::from_u8(6 - i).unwrap());
+        }
+    }
+    operating_days.sort_by_key(|w| w.num_days_from_monday());
+}
+
+fn clean_extra_dates(start_date: Date, end_date: Date, dates: &mut Vec<ExceptionDate>) {
+    dates.retain(|d| d.date >= start_date && d.date <= end_date);
+}
+///Allows you to present a list of dates in a readable way.
+pub fn translate(dates: &BTreeSet<Date>) -> BlockPattern {
+    let mut res = BlockPattern::default();
+    let start_date = match dates.iter().next() {
+        Some(d) => *d,
+        None => return res,
+    };
+    let end_date = match dates.iter().next_back() {
+        Some(d) => *d,
+        None => return res,
+    };
+
+    let validity_period = ValidityPeriod {
+        start_date: start_date,
+        end_date: end_date,
+    };
+    res.validity_periods.push(validity_period);
+
+    let mut monday_ref = get_prev_monday(start_date);
+    let validity_pattern = compute_validity_pattern(monday_ref, end_date, &dates);
+    let best_week = get_min_week_pattern(&validity_pattern);
+    fill_operating_days(best_week, &mut res.operating_days);
+
+    for week in validity_pattern {
+        if week != best_week {
+            let exception: u8 = (!best_week) & week;
+            fill_exceptions(
+                monday_ref,
+                exception,
+                ExceptionType::Add,
+                &mut res.exceptions,
+            );
+
+            let exception: u8 = (week ^ best_week) & best_week;
+            fill_exceptions(
+                monday_ref,
+                exception,
+                ExceptionType::Remove,
+                &mut res.exceptions,
+            );
+        }
+        monday_ref += Duration::days(7);
+    }
+    clean_extra_dates(start_date, end_date, &mut res.exceptions);
+    res
+}
+
+//       July 2012
+// Mo Tu We Th Fr Sa Su
+//                    1
+//  2  3  4  5  6  7  8
+//  9 10 11 12 13 14 15
+// 16 17 18 19 20 21 22
+// 23 24 25 26 27 28 29
+// 30 31
+
+#[test]
+fn nb_weeks() {
+    let mut dates = BTreeSet::new();
+    let mut start_date: Date;
+    let mut end_date: Date;
+
+    start_date = Date::from_ymd(2012, 7, 2);
+    assert_eq!(get_prev_monday(start_date), start_date);
+
+    start_date = Date::from_ymd(2012, 7, 5);
+    assert_eq!(get_prev_monday(start_date), Date::from_ymd(2012, 7, 2));
+
+    start_date = Date::from_ymd(2012, 7, 8);
+    assert_eq!(get_prev_monday(start_date), Date::from_ymd(2012, 7, 2));
+
+    // one week
+    start_date = Date::from_ymd(2012, 7, 2);
+    end_date = Date::from_ymd(2012, 7, 8);
+    dates.insert(start_date);
+    dates.insert(end_date);
+    assert_eq!(
+        compute_validity_pattern(get_prev_monday(start_date), end_date, &dates).len(),
+        1
+    );
+
+    // partial + one week
+    dates.clear();
```

Comment:
It's better to split this in several test, this way you don't have to cleanup, and you can have several fails at once (else it will stop at the first one)